_source/_posts/2022-01-13-spring-cloud-contract.md

```diff
@@ -0,0 +1,893 @@
+
+@RestController
+public class HatController {
+    private final HatService hatService;
+
+    public HatController(HatService hatService) {
+        this.hatService = hatService;
+    }
+
+    @GetMapping("hat/{id}")
+    public Hat findHatById(@PathVariable("id") Long id) {
+        Hat hat = hatService.findHatById(id);
+        if (hat == null) {
+            throw new ResponseStatusException(
+                    HttpStatus.NOT_FOUND, "Not Found"
+            );
+        }
+        return hat;
+    }
+}
+```
+
+The HatController uses Spring's dependency injection to get the HatService as a bean. It then uses the service to lookup hats by ID and serve them to the consumer of the REST service endpoint at `/hat/{id}`. 
+
+You now have a working hat producer. Start the project with the following command:
+
+```bash
+./mvnw spring-boot:run
+```
+
+Once the service has started, you can open a second terminal and use HTTPie to test it.
+
+```bash
+http :8080/hat/1
+```
+
+```bash
+HTTP/1.1 200 
+...
+{
+    "color": "red",
+    "id": 1,
+    "name": "Sombrero",
+    "size": 30
+}
+```
+
+## Create the Producer Contract
+
+Next, you need to write the contracts that define the interservice agreement to govern the behavior of the producer. This is where Spring Cloud Contract comes in and things get interesting.
+
+You're going to create three contracts under `src/test/resources/contracts`. The contracts can be written in YAML or as a Groovy, Java, or Kotlin DSL script. In this tutorial I've chosen to go with the Groovy. According to the spec, you can put multiple contracts in one file, which is what I initially did. This worked fine on the producer side, but when I tried to import the stubs on the consumer side for integration testing, it didn't work. For this reason, the three contracts below are in three separate files. 
+
+Create the first contract in your `src/test/resources` directory. 
+
+`src/test/resources/contracts/find_hat_by_id.groovy`
+
+```groovy
+package contracts
+
+import org.springframework.cloud.contract.spec.Contract
+
+Contract.make {
+    description "should return hat by id=1"
+
+    request {
+        url "/hat/1"
+        method GET()
+    }
+
+    response {
+        status OK()
+        headers {
+            contentType applicationJson()
+        }
+        body (
+                id: 1,
+                name: "Test Hat 1",
+                size: 10,
+                color: "striped"
+        )
+    }
+
+}
+```
+
+In the `request` block, this contract defines the request method at `GET` and the request URL as `/hat/1`. In the `response` block, this contract specifies that the status should be `OK`, the headers should be `applicationJson`, and the attributes of the red sombrero should be returned. This contract is pretty simple, but the DSL has a lot of power. [Take a look at the docs](https://cloud.spring.io/spring-cloud-contract/reference/html/project-features.html) to dig deeper.
```

Comment:
"this contract defines the request method at `GET`" -> 
this contract defines the request method as `GET`
at-> as ?
 is this a typo? Not sure.